test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+// to use. testWriteToOutputArray expects that the large array that it receives is
+// equal to the same large array with the gold array injected on the same subarray location
+void genSubArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    const dim_t pad_size = 2;
+
+    // The large array is padded on both sides of each dimension
+    // Padding is only applied if the dimension is used, i.e. if dims[i] > 1
+    af::dim4 full_arr_dims(dims[0] > 1 ? dims[0] + 2*pad_size : dims[0],
+                           dims[1] > 1 ? dims[1] + 2*pad_size : dims[1],
+                           dims[2] > 1 ? dims[2] + 2*pad_size : dims[2],
+                           dims[3] > 1 ? dims[3] + 2*pad_size : dims[3]);
+
+    // Calculate index of sub-array. These will be used also by
+    // testWriteToOutputArray so that the gold sub array will be placed in the
+    // same location. Currently, this location is the center of the large array
+    af::index subarr_idxs[4];
+    for (int i = 0; i < 4; ++i) {
+        subarr_idxs[i] =
+            dims[i] > 1 ? af::seq(pad_size, pad_size + dims[i] - 1) : af::span;
+    }
+
+    metadata.initArrays(full_arr_dims, ty, subarr_idxs);
+}
+
+// Generates a reordered array. testWriteToOutputArray expects that this array
+// will still have the correct output values from the af_* function, even though
+// the array was initially reordered.
+void genReorderedArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    // This reorder combination will not move data around, but will simply
+    // call modDims and modStrides (see src/api/c/reorder.cpp).
+    // The output will be checked if it is still correct even with the
+    // modified dims and strides "hack" with no data movement
+    unsigned reorder_0 = 0;
+    unsigned reorder_1 = 2;
+    unsigned reorder_2 = 1;
+
+    // Shape the output array such that the reordered output array will have
+    // the correct dimensions that the test asks for (i.e. must match dims arg)
+    af::dim4 out_dims(dims[reorder_0], dims[reorder_1], dims[reorder_2]);
+    metadata.initArrays(out_dims, ty);
+    metadata.setOutputArray(af::reorder(metadata.out_arr,
+                                        reorder_0, reorder_1, reorder_2));
+}
+
+// Partner function of testWriteToOutputArray. This generates the "special"
+// array that testWriteToOutputArray will use to check if the af_* function
+// correctly uses an existing array as its output
+void genTestOutputArray(af_array *out_ptr,
+                        const unsigned ndims, const dim_t *dims,
+                        const af::dtype ty,
+                        TestOutputArrayInfo* metadata) {
+    af::dim4 arr_dims(ndims, dims);
+    switch (metadata->getOutputArrayType()) {
+    case NULL_ARRAY:
+        genNullArray(*metadata, arr_dims, ty);
+        break;
```

Comment:
I would make null array case as the default case, that way, the whole function `genNullArray` can be removed as it won't do anything in the modified code.
```C++
switch(...) {
...
default: // do nothing as that itself is null array case
}
```

Reply:
What if we make sure we give back 0 for the `af_array` in this case (this is what I currently do inside `genNullArray`)? Maybe I'm a little paranoid, but if the user doesn't initialize the output `af_array` (passed in here as `*out_ptr`) before calling `genTestOutputArray`, then I'd guess that it could be something other than 0 and we'd get undefined/ugly behavior.